core/src/overlay.rs

```diff
@@ -0,0 +1,423 @@
+}
+
+enum OverlayInnerState {
+  ToShow(Instant, Rc<Window>),
+  Showing(WidgetId, Rc<Window>),
+  Hided,
+}
+
+#[derive(Clone)]
+struct OverlayState(Rc<RefCell<OverlayInnerState>>);
+impl Default for OverlayState {
+  fn default() -> Self { OverlayState(Rc::new(RefCell::new(OverlayInnerState::Hided))) }
+}
+
+impl OverlayState {
+  fn remove(&self) {
+    let state = replace(&mut *self.0.borrow_mut(), OverlayInnerState::Hided);
+    if let OverlayInnerState::Showing(wid, wnd) = state {
+      let _ = AppCtx::spawn_local(async move {
+        let root = wnd.widget_tree.borrow().root();
+        let build_ctx = BuildCtx::new(None, &wnd.widget_tree);
+        build_ctx.dispose_subtree(wid);
+        build_ctx.mark_dirty(root);
+      });
+    }
+  }
+
+  fn is_show(&self) -> bool { !matches!(*self.0.borrow(), OverlayInnerState::Hided) }
+
+  fn show(&self, w: impl WidgetBuilder + 'static, style: OverlayStyle, wnd: Rc<Window>) {
+    if self.is_show() {
+      return;
+    }
+    let this = self.clone();
+    let instant = Instant::now();
+    *this.0.borrow_mut() = OverlayInnerState::ToShow(instant, wnd);
+    let _ = AppCtx::spawn_local(async move {
+      let wnd = match (instant, &*this.0.borrow()) {
+        (instant, OverlayInnerState::ToShow(crate_at, wnd)) if &instant == crate_at => wnd.clone(),
+        _ => return,
+      };
+      let build_ctx = BuildCtx::new(None, &wnd.widget_tree);
+      let w = this.wrap_style(w, style).widget_build(&build_ctx);
+      let wid = w.id();
+      *this.0.borrow_mut() = OverlayInnerState::Showing(wid, wnd.clone());
+      let root = wnd.widget_tree.borrow().root();
+      build_ctx.append_child(root, w);
+      build_ctx.on_subtree_mounted(wid);
+      build_ctx.mark_dirty(wid);
+    });
+  }
+
+  fn wrap_style(&self, w: impl WidgetBuilder, style: OverlayStyle) -> impl WidgetBuilder {
+    let this = self.clone();
+    fn_widget! {
+      let OverlayStyle { close_policy, mask_brush } = style;
+      let this2 = this.clone();
+      @Container {
+        size: Size::new(f32::INFINITY, f32::INFINITY),
+        background: mask_brush.unwrap_or_else(|| Color::from_u32(0).into()),
+        on_tap: move |e| {
+          if close_policy.contains(ClosePolicy::TAP_OUTSIDE)
+            && e.target() == e.current_target() {
+            this.remove();
+          }
+        },
+        on_key_down: move |e| {
+          if close_policy.contains(ClosePolicy::ESC)
+            && *e.key() == VirtualKey::Named(NamedKey::Escape) {
+            this2.remove();
+          }
+        },
+        @$w {}
+      }
+    }
+  }
+
+  fn close_handle(&self) -> OverlayCloseHandle { OverlayCloseHandle(self.clone()) }
+}
+
+impl<O, M> OverlayBuilder for Overlay<M>
+where
+  M: Fn(OverlayCloseHandle) -> O + 'static,
+  O: WidgetBuilder + 'static,
+{
+  type Output = O;
+  fn build(&self, ctrl: OverlayCloseHandle) -> Self::Output { (self.0.builder)(ctrl) }
+}
+
+#[derive(Clone)]
+pub struct OverlayWidget<M>(M);
```

Comment:
The inner type does not need to be public

Reply:
the type is refed by a pub method from_widget as return type